pkg/statsd/cloud_handler.go

```diff
@@ -138,19 +141,29 @@ func (ch *CloudHandler) Run(ctx context.Context) error {
 		select {
 		case <-ctx.Done():
 			return ctx.Err()
+		case toLookupC <- toLookupIP:
+			toLookupIP = gostatsd.UnknownIP
+			toLookupC = nil // ip has been sent; if there is nothing to send, will block
 		case lr := <-lookupResults:
-			ch.handleLookupResult(ctx, lr, awaitingMetrics, awaitingEvents)
+			ch.handleLookupResult(ctx, lr)
 		case t := <-refreshTicker.C:
-			ch.doRefresh(ctx, toLookup, t)
+			ch.doRefresh(ctx, t)
 		case m := <-ch.metricSource:
-			ch.handleMetric(ctx, toLookup, m, awaitingMetrics)
+			ch.handleMetric(ctx, m)
 		case e := <-ch.eventSource:
-			ch.handleEvent(ctx, toLookup, e, awaitingEvents)
+			ch.handleEvent(ctx, e)
+		}
+		if toLookupC == nil && len(ch.toLookupIPs) > 0 {
+			last := len(ch.toLookupIPs) - 1
+			toLookupIP = ch.toLookupIPs[last]
+			ch.toLookupIPs[last] = gostatsd.UnknownIP // Enable GC
+			ch.toLookupIPs = ch.toLookupIPs[:last]
```

Comment:
Ideally this should be a ring buffer to preserve priority. I was hesitant to do `ch.toLookupIPs[1:]` to avoid copying the whole slice contents.